src/main/java/org/opensearch/knn/index/memory/NativeMemoryAllocation.java

```diff
@@ -163,11 +164,19 @@ class IndexAllocation implements NativeMemoryAllocation {
 
         @Override
         public void close() {
-            executor.execute(() -> {
+            Runnable onClose = () -> {
                 writeLock();
                 cleanup();
                 writeUnlock();
-            });
+            };
+
+            // The close operation needs to be blocking to prevent overflow
+            // This blocks any entry until the close has completed, preventing creation before close scenarios
+            if (KNNFeatureFlags.isForceEvictCacheEnabled()) {
```

Comment:
Have you tested with only this change without a code which try to evict before adding value to cache to see if node crash or not?

Reply:
Yes. As I pointed out in the testing section - the nodes did crash.